### PR TITLE
Added support to video streaming using a UVC DEVICE + Translation to Brazilian Portuguese (PT-BR)

### PR DESCRIPTION
**Pull request by guiseco**
Added Feature to insert FPV on 3DR Tower.

https://goo.gl/photos/EbPe6FteiWYeEmXr9

https://goo.gl/photos/BJH7HoPE8uBnoLJo9

5aa64ace-8952-11e5-8aad-e4511870156a
d945f562-8927-11e5-9616-4199540da146
e66dfd5c-8927-11e5-97e0-f34ec572a786
e66f7b32-8927-11e5-9172-7d67b16726a6
e66f709c-8927-11e5-9a45-8904184bb93f
e672a5e6-8927-11e5-8763-8d63896798b7

### CODE REVIEW

**Comment by hydropik** [comment]
Hello, 
Good work I hope the support of UVC will be added to the public release.
Thank you.

**Comment by ne0fhyk** [comment]
@guiseco sorry for the delay in reviewing this pr. I'll review it tonight, and add feedback as needed!

**Comment by ne0fhyk** [comment]
@guiseco while I'm reviewing the pr, could you update the [Tower wiki](https://github.com/DroidPlanner/Tower/wiki) with instructions to setup and use video streaming via a uvc device.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+import com.serenegiant.usb.USBMonitor;
+
+import org.droidplanner.android.R;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+
+public class UVCDialog extends DialogFragment {
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco can you make use of the `DialogFragment` from the support library instead.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+    /**
+     * Helper method
+     * @param parent FragmentActivity
+     * @return
+     */
+    public static UVCDialog showDialog(final Activity parent) {
+        UVCDialog dialog = newInstance();
+        try {
+            dialog.show(parent.getFragmentManager(), TAG);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco here as well, let's use the support library equivalent: `getSupportFragmentManager()`.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * Helper method
+     * @param parent FragmentActivity
+     * @return
+     */
+    public static UVCDialog showDialog(final Activity parent, USBMonitor mUSBMonitor) {
+        UVCDialog dialog = newInstance();
+        dialog.mUSBMonitor = mUSBMonitor;
+        try {
+            dialog.show(parent.getFragmentManager(), TAG);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco here as well, let's use the support library equivalent: `getSupportFragmentManager()`.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        } catch (final IllegalStateException e) {
+            dialog = null;
+        }
+        return dialog;
+    }
+
+    public static UVCDialog newInstance() {
+        final UVCDialog dialog = new UVCDialog();
+        final Bundle args = new Bundle();
+        dialog.setArguments(args);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco what's the `Bundle args` for here, since no arguments is being added to it?

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+    @Override
+    public void onAttach(final Activity activity) {
+        super.onAttach(activity);
+    }
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (savedInstanceState == null)
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco not sure what you're trying to do here.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            savedInstanceState = getArguments();
+    }
+
+    @Override
+    public void onSaveInstanceState(final Bundle saveInstanceState) {
+        final Bundle args = getArguments();
+        if (args != null) saveInstanceState.putAll(args);
+        super.onSaveInstanceState(saveInstanceState);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco same here. I'm not sure what's the goal here.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        mSpinner.setEmptyView(empty);
+        return rootView;
+    }
+
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        updateDevices();
+        final Button button = (Button)getDialog().findViewById(android.R.id.button3);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco this should be within the `initView()` method instead.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    };
+
+    private final DialogInterface.OnClickListener mOnDialogClickListener = new DialogInterface.OnClickListener() {
+        @Override
+        public void onClick(final DialogInterface dialog, final int which) {
+            switch (which) {
+                case DialogInterface.BUTTON_POSITIVE:
+                    final Object item = mSpinner.getSelectedItem();
+                    if (item instanceof UsbDevice) {
+                        mUSBMonitor.requestPermission((UsbDevice)item);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco check `mUSBMonitor` is non null before dereferencing it.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                    }
+                    break;
+            }
+        }
+    };
+
+    public void updateDevices() {
+        mUSBMonitor.dumpDevices();
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco here as well, let's check `mUSBMonitor` is non null before dereferencing it.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        final List<DeviceFilter> filter = DeviceFilter.getDeviceFilters(getActivity(), R.xml.uvc_device_filter);
+        mDeviceListAdapter = new DeviceListAdapter(getActivity(), mUSBMonitor.getDeviceList(filter.get(0)));
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco in addition of the null check, let's make sure `filter` is not empty before retrieving the first element.

File: Android/src/org/droidplanner/android/dialogs/UVCDialog.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        }
+
+        @Override
+        public int getCount() {
+            return mList.size();
+        }
+
+        @Override
+        public UsbDevice getItem(final int position) {
+            if ((position >= 0) && (position < mList.size()))
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/dialogs/UVCDialog.java]: @guiseco you don't need to check the validity of the `position` argument here, since it's based on the `getCount()` that you overrode above.

File: Android/src/org/droidplanner/android/fragments/widget/video/BaseUVCVideoWidget.kt [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    protected var mUVCCamera: UVCCamera? = null
+    protected var isPreview:Boolean = false
+    protected var usbDevice: UsbDevice? = null
+    protected var mPreviewSurface: Surface? = null
+
+    //Aspect ratio
+    protected var aspectRatio_4_3: Float = 3f / 4f
+    protected var aspectRatio_16_9: Float = 9f / 16f
+    protected var aspectRatio_21_9: Float = 9f / 21f
+    protected var aspectRatio_1_1: Float = 1f / 1f
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/fragments/widget/video/BaseUVCVideoWidget.kt]: @guiseco seems like the aspect ratios above should be defined as constants.

File: Android/src/org/droidplanner/android/fragments/widget/video/BaseUVCVideoWidget.kt [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    protected val receiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context, intent: Intent) {
+            when (intent.action) {
+                AttributeEvent.STATE_CONNECTED -> {
+                    startVideoStreaming()
+                }
+                AttributeEvent.STATE_ARMING -> {
+                    val droneState = drone.getAttribute<State>(AttributeType.STATE)
+                    if (droneState.isArmed){
+                        startVideoStreaming()
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/fragments/widget/video/BaseUVCVideoWidget.kt]: @guiseco since you're streaming on `STATE_CONNECTED` above, I'd assume you'd also want to stream here regardless of the arming status?

File: Android/src/org/droidplanner/android/fragments/widget/video/FullWidgetUVCLinkVideo.kt [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        takePhotoButton?.setOnClickListener {
+            Toast.makeText(context, "NOT implemented YET", Toast.LENGTH_SHORT).show()
+        }
+
+        recordVideo?.setOnClickListener {
+            Toast.makeText(context, "NOT implemented YET", Toast.LENGTH_SHORT).show()
+        }
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/fragments/widget/video/FullWidgetUVCLinkVideo.kt]: @guiseco should the `takePhoto` and `recordVideo` buttons be hidden (visibility set to `GONE`) until the feature is implemented?

File: Android/src/org/droidplanner/android/notifications/TTSNotificationProvider.java [hunk 15]
```diff
@@ -420,7 +420,7 @@ public void onInit(int status) {
     }
 
     private void speak(String string) {
-        speak(string, false, null);
+        speak(string, true, null);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/notifications/TTSNotificationProvider.java]: @guiseco why the change?

File: Android/src/org/droidplanner/android/utils/prefs/DroidPlannerPrefs.java [hunk 16]
```diff
@@ -502,4 +504,12 @@ public void setCustomVideoUdpPort(int udpPort){
     public int getCustomVideoUdpPort(){
         return prefs.getInt(PREF_CUSTOM_VIDEO_UDP_PORT, -1);
     }
+
+    public void setUVCVideoAspectRatio(Float aspectRatio){
+        prefs.edit().putFloat(PREF_UVC_VIDEO_ASPECT_RATIO, aspectRatio).apply();
+    }
+
+    public Float getUVCVideoAspectRatio(){
+        return prefs.getFloat(PREF_UVC_VIDEO_ASPECT_RATIO, 3f / 4f);
```
Comment (ne0fhyk) [inline Android/src/org/droidplanner/android/utils/prefs/DroidPlannerPrefs.java]: @guiseco let's turn `3f / 4f` into constant.

File: Android/src/org/droidplanner/android/notifications/TTSNotificationProvider.java (hunk 15 above)
Comment (guiseco) [inline Android/src/org/droidplanner/android/notifications/TTSNotificationProvider.java]: If false, speech is overlapped by another speech

**Comment by guiseco** [comment]
@ne0fhyk Could you check again ?

**Comment by ne0fhyk** [comment]
@guiseco I'm on it!

**Comment by ne0fhyk** [comment]
Looks good, merging!
